Mathlib/RingTheory/HahnSeries/Lex.lean

```diff
@@ -0,0 +1,170 @@
+  le_total a b := by
+    rcases eq_or_ne a b with hab | hab
+    · exact Or.inl hab.le
+    · have hab := Function.ne_iff.mp <| HahnSeries.ext_iff.ne.mp hab
+      let u := {i : Γ | (ofLex a).coeff i ≠ 0} ∪ {i : Γ | (ofLex b).coeff i ≠ 0}
+      let v := {i : Γ | (ofLex a).coeff i ≠ (ofLex b).coeff i}
+      have hvu : v ⊆ u := by
+        intro i h
+        rw [Set.mem_union, Set.mem_setOf_eq, Set.mem_setOf_eq]
+        contrapose! h
+        rw [Set.notMem_setOf_iff, Mathlib.Tactic.PushNeg.not_ne_eq, h.1, h.2]
+      have hv : v.IsWF :=
+        ((ofLex a).isPWO_support'.isWF.union (ofLex b).isPWO_support'.isWF).subset hvu
+      let i := hv.min hab
+      have hji (j) : j < i → (ofLex a).coeff j = (ofLex b).coeff j :=
+        not_imp_not.mp <| fun h' ↦ hv.not_lt_min hab h'
+      have hne : (ofLex a).coeff i ≠ (ofLex b).coeff i := hv.min_mem hab
+      obtain hi | hi := lt_or_gt_of_ne hne
+      · exact Or.inl (le_of_lt ⟨i, hji, hi⟩)
+      · exact Or.inr (le_of_lt ⟨i, fun j hj ↦ (hji j hj).symm, hi⟩)
+  toDecidableLE := Classical.decRel _
+
+theorem leadingCoeff_pos_iff {x : Lex (HahnSeries Γ R)} : 0 < (ofLex x).leadingCoeff ↔ 0 < x := by
+  rw [lt_iff]
+  constructor
+  · intro hpos
+    have hne : (ofLex x) ≠ 0 := leadingCoeff_ne_iff.mp hpos.ne.symm
+    have htop : (ofLex x).orderTop ≠ ⊤ := ne_zero_iff_orderTop.mp hne
+    use (ofLex x).orderTop.untop htop
+    constructor
+    · intro j hj
+      simpa using (coeff_eq_zero_of_lt_orderTop ((WithTop.lt_untop_iff htop).mp hj)).symm
+    · rw [coeff_untop_eq_leadingCoeff hne]
+      simpa using hpos
+  · intro ⟨i, hj, hi⟩
+    have horder : (ofLex x).orderTop = WithTop.some i := by
+      apply orderTop_eq_of_le
+      · simpa using hi.ne.symm
+      · intro g hg
+        contrapose! hg
+        simpa using (hj g hg).symm
+    have htop : (ofLex x).orderTop ≠ ⊤ := WithTop.ne_top_iff_exists.mpr ⟨i, horder.symm⟩
+    have hne : ofLex x ≠ 0 := ne_zero_iff_orderTop.mpr htop
+    have horder' : (ofLex x).orderTop.untop htop = i := (WithTop.untop_eq_iff _).mpr horder
+    rw [← coeff_untop_eq_leadingCoeff hne, horder']
+    simpa using hi
+
+theorem leadingCoeff_nonneg_iff {x : Lex (HahnSeries Γ R)} :
+    0 ≤ (ofLex x).leadingCoeff ↔ 0 ≤ x := by
+  constructor
+  · intro h
+    obtain heq | hlt := h.eq_or_lt
+    · exact le_of_eq (leadingCoeff_eq_iff.mp heq.symm).symm
+    · exact (leadingCoeff_pos_iff.mp hlt).le
+  · intro h
+    obtain rfl | hlt := h.eq_or_lt
+    · simp
+    · exact (leadingCoeff_pos_iff.mpr hlt).le
+
+theorem leadingCoeff_neg_iff {x : Lex (HahnSeries Γ R)} : (ofLex x).leadingCoeff < 0 ↔ x < 0 := by
+  simpa using (leadingCoeff_nonneg_iff (x := x)).not
+
+theorem leadingCoeff_nonpos_iff {x : Lex (HahnSeries Γ R)} :
+    (ofLex x).leadingCoeff ≤ 0 ↔ x ≤ 0 := by
+  simpa using (leadingCoeff_pos_iff (x := x)).not
+
+end LinearOrder
+
+section OrderedMonoid
+
+variable [PartialOrder R] [AddCommMonoid R] [AddLeftStrictMono R] [IsOrderedAddMonoid R]
```

Comment:
nit: I think we prefer
```suggestion
section OrderedMonoid
variable [PartialOrder R] [AddCommMonoid R] [AddLeftStrictMono R] [IsOrderedAddMonoid R]
```

which makes it easier to see what the section is about. The blank lines was an accidental style change induced by an automated tool we no longer use.